src/dataflow-types/src/client/controller.rs

```diff
@@ -367,75 +373,115 @@ impl<T> Controller<T>
 where
     T: Timestamp + Lattice + Codec64,
 {
-    pub async fn recv(&mut self) -> Result<Option<ControllerResponse<T>>, anyhow::Error> {
-        let mut storage_alive = true;
-        loop {
+    /// Wait until the controller is ready to process a response.
+    ///
+    /// This method may await indefinitely.
+    ///
+    /// This method is intended to be cancel-safe: dropping the returned
+    /// future at any await point and restarting from the beginning is fine.
+    // This method's correctness relies on the assumption that the _underlying_
+    // clients are _also_ cancel-safe, since it introduces its own `select` call.
```

Comment:
sorry to bring this up on a merged pr, but just curious, how did we determine that the storage-controller was cancel safe?

Reply:
Because all it does is await receiving from a GenericClient, so it follows from the hypothesis that all clients are now safe. (Which I believe to be true, unless there are still bugs)